abdm_integrator/hiu/views/consents.py

```diff
@@ -102,30 +102,39 @@ class GatewayConsentRequestNotifyProcessor:
     def __init__(self, request_data):
         self.request_data = request_data
 
-    @transaction.atomic
     def process_request(self):
         consent_status = self.request_data['notification']['status']
         if consent_status == ConsentStatus.REVOKED:
             return self.handle_revoked()
         consent_request = ConsentRequest.objects.get(
             consent_request_id=self.request_data['notification']['consentRequestId']
         )
-        consent_request.update_status(consent_status)
         if consent_status == ConsentStatus.GRANTED:
             self.handle_granted(consent_request)
         elif consent_status == ConsentStatus.EXPIRED:
             self.handle_expired(consent_request)
 
     def handle_granted(self, consent_request):
-        for artefact in self.request_data['notification']['consentArtefacts']:
-            consent_artefact = ConsentArtefact(artefact_id=artefact['id'], consent_request=consent_request)
-            self.artefact_fetch(consent_artefact)
+        consent_artefacts = []
+        with transaction.atomic():
+            consent_request.update_status(ConsentStatus.GRANTED)
+            for artefact in self.request_data['notification']['consentArtefacts']:
+                consent_artefact, _ = ConsentArtefact.objects.get_or_create(artefact_id=artefact['id'],
+                                                                         consent_request=consent_request)
+                consent_artefacts.append(consent_artefact)
+
+        for consent_artefact in consent_artefacts:
+            if consent_artefact.fetch_status == ArtefactFetchStatus.PENDING:
+                self.artefact_fetch(consent_artefact)
```

Comment:
We have to be careful with the naming. It's hard to understand that `artefact_fetch` is going to update the `consent_artefact` object itself. If the methods have unexepcted side effects the names should indicate that it so.

Reply:
Noted. Will keep this in mind.
Addressed in 5199acecb6c64767fe7fee018161c80151346ac5